service/src/BatchServer.cpp

```diff
@@ -144,42 +162,85 @@ namespace geopm
             }
             catch (const Exception &ex) {
                 if (ex.err_value() != ESRCH) {
-                    throw ex;
+                    throw;
                 }
             }
             m_is_active = false;
         }
     }
 
+    char BatchServerImp::read_message(void)
+    {
+        char in_message;
+        try {
+            in_message = m_batch_status->receive_message();
+        }
+        catch (const Exception &ex) {
+            // If we were interupted by SIGTERM exit gracefully
+            if (ex.err_value() == EINTR &&
+                g_sigterm_count != 0) {
+                in_message = BatchStatus::M_MESSAGE_TERMINATE;
+            }
+            else {
+                throw;
+            }
+        }
+        return in_message;
+    }
+
+    void BatchServerImp::write_message(char out_message)
+    {
+        if (m_is_client_waiting) {
+            try {
+                m_batch_status->send_message(out_message);
+                m_is_client_waiting = false;
+            }
+            catch (const Exception &ex) {
+                // If we were not interupted SIGTERM rethrow
+                if (ex.err_value() != EINTR ||
```

Comment:
Why are we not setting the `in_message = BatchStatus::M_MESSAGE_TERMINATE;` in this situation?

Reply:
I have refactored this code a bit in an attempt to clarify what is going on for the reader.  Let me know if it makes more sense now.